### PR TITLE
feat(html/minifier): allow to specify options for parser/minifier/codegen

### PR DESCRIPTION
**Pull request by alexander-akait**
**Description:**

Allow to specify parser/minifier/codegen options for JS/CSS/JSON

**BREAKING CHANGE:**

Yes

**Related issue (if exists):**

No

### CODE REVIEW

File: crates/swc_css_minifier/src/options.rs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+pub struct MinifyOptions {}
```
Comment (alexander-akait) [inline crates/swc_css_minifier/src/options.rs:6]: For future options

File: crates/swc_ecma_minifier/src/option/mod.rs [hunk 2]
```diff
@@ -28,7 +28,7 @@ pub struct ExtraOptions {
     pub top_level_mark: Mark,
 }
 
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Debug, Serialize, Deserialize, Clone)]
```
Comment (kdy1) [inline crates/swc_ecma_minifier/src/option/mod.rs:31]: I didn't implemented clone intentionall.

File: crates/swc_css_parser/src/parser/mod.rs [hunk 3]
```diff
@@ -18,13 +19,16 @@ mod value;
 
 pub type PResult<T> = Result<T, Error>;
 
-#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[derive(
+    Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize,
+)]
```
Comment (kdy1) [inline crates/swc_css_parser/src/parser/mod.rs:24]: Please add `#[serde(rename_all = "camelCase")]`

File: crates/swc_ecma_codegen/src/config.rs [hunk 4]
```diff
@@ -1,6 +1,7 @@
+use serde::{Deserialize, Serialize};
 use swc_ecma_ast::EsVersion;
 
-#[derive(Debug, Clone, Copy)]
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
```
Comment (kdy1) [inline crates/swc_ecma_codegen/src/config.rs:4]: Please add `#[serde(rename_all = "camelCase")]`

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!

---

swc-bump:
 - swc_css_codegen
 - swc_css_minifier --breaking
 - swc_css_parser
 - swc_ecma_codegen

**Comment by alexander-akait** [comment]
@kdy1 Fixed + refactor to avoid extra clone, unfortenatally we can't avoid `.clone()` for ecma minifier options, because we need to change options based on `type` of the `script` element

File: crates/swc_ecma_codegen/src/config.rs [hunk 5]
```diff
@@ -11,13 +13,16 @@ pub struct Config {
     /// e.g. `const foo = 1;` with [EsVersion::Es3] will emitted as `const foo =
     /// 1` without verification.
     /// This is because it's not a concern of the code generator.
+    #[serde(default)]
```
Comment (kdy1) [inline crates/swc_ecma_codegen/src/config.rs:16]: This should be `EsVersion::latest()`  by default

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc_ecma_codegen/src/config.rs (hunk 5 above)
Comment (alexander-akait) [reply crates/swc_ecma_codegen/src/config.rs]: @kdy1 Fixed :+1:

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_css_codegen
 - swc_css_minifier
 - swc_css_parser
 - swc_ecma_codegen
 - swc_ecma_minifier
 - swc_html_minifier --breaking

**Pull request merged**